src/server/server_family.cc

```diff
@@ -1069,6 +1070,52 @@ void ServerFamily::Client(CmdArgList args, ConnectionContext* cntx) {
   return (*cntx)->SendError(UnknownSubCmd(sub_cmd, "CLIENT"), kSyntaxErrType);
 }
 
+void ServerFamily::Cluster(CmdArgList args, ConnectionContext* cntx) {
+  ToUpper(&args[1]);
+  string_view sub_cmd = ArgS(args, 1);
+
+  if (dfly::is_emulated_cluster == false) {
+    return (*cntx)->SendError("CLUSTER commands demands the `cluster_mode` flag set to `emulated`");
+  }
+
+  if (sub_cmd == "HELP") {
+    string_view help_arr[] = {
+        "CLUSTER <subcommand> [<arg> [value] [opt] ...]. Subcommands are:",
+        "SLOTS",
+        "    Returns details about which cluster slots map to which instances.",
+        "HELP",
+        "    Prints this help.",
+    };
+    return (*cntx)->SendSimpleStrArr(help_arr, ABSL_ARRAYSIZE(help_arr));
+  }
+
+  if (sub_cmd == "SLOTS") {
+    (*cntx)->StartArray(1);    // Replying just one cluster
+    (*cntx)->StartArray(3);    // Should return a replica with master info?
+    (*cntx)->SendLong(0);      // start slot
+    (*cntx)->SendLong(16383);  // end slot
+    string cluster_announce_ip = GetFlag(FLAGS_cluster_announce_ip);
+    string_view ip_addr = cluster_announce_ip.empty()
```

Comment:
when I run redis6 with `--cluster-enabled yes` I see that "cluster slots" return empty string if `cluster_announce_ip` is not defined.

Reply:
The problem here is that redis-py is using len() on this value, and if I use `(*cntx)->SendNull("")` it throws an error.
If I use `(*cntx)->SendSendSimpleString("")` or `(*cntx)->SendBulkString("")` redis-py doesn't complain.

Reply:
I've checked the Redis sources, it returns null indeed https://github.com/redis/redis/blob/203b12e41ff7981f0fae5b23819f072d61594813/src/cluster.c#L5311

You can leave it as a to-do

Reply:
It seems they have two different 'Null' returns https://github.com/redis/redis/blob/fac188b49d9680fdeb90332f163e5634cec1ea13/src/networking.c#L986 and dragonfly has just one https://github.com/dragonflydb/dragonfly/blob/0377481e28d776e434019cec7c5b98057b943326/src/facade/reply_builder.cc#L198 that is probably the difference here.

Reply:
Indeed, RESP3 added a new null type
 
https://github.com/redis/redis-specifications/blob/master/protocol/RESP3.md#:~:text=Number%20type%20instead.-,Null,-The%20null%20type

Reply:
The message formatting was fixed, and redis-cli, python and JS all working with the reply from DF